internal/controller/accessrequest_controller.go

```diff
@@ -33,30 +38,241 @@ type AccessRequestReconciler struct {
-// the user.
 //
-// For more details, check Reconcile and its Result here:
-// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.18.2/pkg/reconcile
+// 1. verify if accessrequest is expired and status is "granted"
+// 1.1 if so, remove the user from the elevated role
+// 1.2 update the accessrequest status to "expired"
+// 1.3 return
+// 2. verify if user has the necessary access to be promoted
+// 2.1 if they don't, update the accessrequest status to "denied"
+// 2.2 return
+// 3. verify if CR is approved
+// 4. retrieve the Application
+// 5. retrieve the AppProject
+// 6. assign user in the desired role in the AppProject
+// 7. update the accessrequest status to "granted"
+// 8. set the RequeueAfter in Result
 func (r *AccessRequestReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = log.FromContext(ctx)
+	logger := log.FromContext(ctx).
+		WithValues("controller", "AccessRequest", "reconcile", req.NamespacedName)
+
+	logger.Info("retrieving AccessRequest k8s state")
+	var ar api.AccessRequest
+	if err := r.Get(ctx, req.NamespacedName, &ar); err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		logger.Error(err, "error retrieving AccessRequest from k8s")
+		return ctrl.Result{}, err
+	}
+
+	// check if the object is being deleted and properly handle it
+	deleted, err := r.handleFinalizer(ctx, &ar)
+	if err != nil {
+		logger.Error(err, "handleFinalizer error")
+		return ctrl.Result{}, fmt.Errorf("error handling finalizer: %w", err)
+	}
+	// stop the reconciliation as the object was deleted
+	if deleted {
+		logger.Info("Object deleted")
+		return ctrl.Result{}, nil
+	}
+
+	// initialize the status if not done yet
+	if ar.Status.Status == "" {
+		ar.UpdateStatus(api.RequestedStatus, "")
+		r.Status().Update(ctx, &ar)
+	}
+
+	// check if the access is expired
+	expired, err := r.handleAccessExpired(ctx, &ar)
+	if err != nil {
+		logger.Error(err, "handleAccessExpired error")
+		return ctrl.Result{}, fmt.Errorf("error handling access expired: %w", err)
+	}
+	// Stop the reconciliation if the access is expired
+	if expired {
+		logger.Info("AccessRequest expired")
+		return ctrl.Result{}, nil
+	}
+
+	// check subject is sudoer
+	status, err := r.handlePermission(ctx, &ar)
+	if err != nil {
+		logger.Error(err, "handlePermission error")
+		return ctrl.Result{}, fmt.Errorf("error handling permission: %w", err)
+	}
+
+	return buildResult(status, &ar), nil
+}
+
+// buildResult will verify the given status and determine when this access
+// request should be requeued.
+func buildResult(status api.Status, ar *api.AccessRequest) ctrl.Result {
+	result := ctrl.Result{}
+	switch status {
+	case api.RequestedStatus:
+		result.Requeue = true
+		// TODO add a controller configuration
+		result.RequeueAfter = time.Minute * 3
+	case api.GrantedStatus:
+		result.Requeue = true
+		result.RequeueAfter = ar.Spec.Duration.Duration
```

Comment:
Should not requeue with a static duration if the goal is to reconcile at a specific time. If the resource is reconciled due to relisting or other k8s event, it will always be delayed by Duration. The value of the requeue should be `(GrantedAt + Duration) - Now`

Reply:
Maybe I am misunderstanding the problem but I think that the current implementation is correct.

Reply:
updated according our conversation